File: apps/android/app/src/test/java/com/foundry/companion/VendorTextTest.kt
package com.foundry.companion

import com.foundry.companion.util.isHiddenVendorText
import com.foundry.companion.util.isIncompleteFunctionCallJson
import com.foundry.companion.util.isVendorFunctionCallPayload
import com.foundry.companion.util.stripVendorToolEcho
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class VendorTextTest {
    private val functionCall = """
        {
          "functionCall": {
            "name": "smith_readiness",
            "args": {
              "operation": "show"
            }
          }
        }
    """.trimIndent()

    @Test
    fun testRecognizesCompleteFunctionCallPayload() {
        assertTrue(isVendorFunctionCallPayload(functionCall))
        assertTrue(isVendorFunctionCallPayload("""{"functionResponse":{"name":"x"}}"""))
        assertFalse(isVendorFunctionCallPayload("Got it."))
        assertFalse(isVendorFunctionCallPayload("""{"ok":true}"""))
    }

    @Test
    fun testHoldsIncompleteFunctionCallJson() {
        assertTrue(isIncompleteFunctionCallJson("{\n  \"functionCall\": {\n    \"name\":"))
        assertFalse(isIncompleteFunctionCallJson(functionCall))
        assertFalse(isIncompleteFunctionCallJson("Got it."))
    }

    @Test
    fun testStripsFinishedEchoes() {
        assertEquals("", stripVendorToolEcho(functionCall))
        assertEquals("", stripVendorToolEcho("Ran `smith_readiness`."))
        assertEquals("Got it.", stripVendorToolEcho("Got it.\nRan `smith_readiness`."))
        assertEquals("Got it.", stripVendorToolEcho("Got it.\n$functionCall"))
    }

    @Test
    fun testHidesEmptyIncompleteAndFinishedEchoes() {
        assertTrue(isHiddenVendorText(""))
        assertTrue(isHiddenVendorText(functionCall))
        assertTrue(isHiddenVendorText("Ran `smith_readiness`."))
        assertTrue(isHiddenVendorText("{\n  \"functionCall\": {"))
        assertFalse(isHiddenVendorText("Got it."))
    }
}

File: apps/desktop/tests/shared/vendor-text.test.ts
/**
 * Pins the vendor-echo filter: Gemini-style functionCall JSON and
 * `Ran \`tool\`.` narration must never reach the operator transcript.
 */

import { describe, expect, it } from 'vitest';
import {
  isHiddenVendorText,
  isIncompleteFunctionCallJson,
  isVendorFunctionCallPayload,
  stripVendorToolEcho,
} from '@shared/vendor-text.js';

const FUNCTION_CALL = `{
  "functionCall": {
    "name": "smith_readiness",
    "args": {
      "operation": "show"
    }
  }
}`;

describe('vendor-text', () => {
  it('recognizes a complete functionCall payload', () => {
    expect(isVendorFunctionCallPayload(FUNCTION_CALL)).toBe(true);
    expect(isVendorFunctionCallPayload('{"functionResponse":{"name":"x"}}')).toBe(true);
    expect(isVendorFunctionCallPayload('Got it.')).toBe(false);
    expect(isVendorFunctionCallPayload('{"ok":true}')).toBe(false);
  });

  it('holds incomplete functionCall JSON until it can parse', () => {
    expect(isIncompleteFunctionCallJson('{\n  "functionCall": {\n    "name":')).toBe(true);
    expect(isIncompleteFunctionCallJson(FUNCTION_CALL)).toBe(false);
    expect(isIncompleteFunctionCallJson('Got it.')).toBe(false);
  });

  it('strips a finished functionCall object and Ran-tool lines', () => {
    expect(stripVendorToolEcho(FUNCTION_CALL)).toBe('');
    expect(stripVendorToolEcho('Ran `smith_readiness`.')).toBe('');
    expect(stripVendorToolEcho('Got it.\nRan `smith_readiness`.')).toBe('Got it.');
    expect(stripVendorToolEcho(`Got it.\n${FUNCTION_CALL}`)).toBe('Got it.');
  });

  it('hides empty, incomplete, and finished vendor echoes', () => {
    expect(isHiddenVendorText('')).toBe(true);
    expect(isHiddenVendorText(FUNCTION_CALL)).toBe(true);
    expect(isHiddenVendorText('Ran `smith_readiness`.')).toBe(true);
    expect(isHiddenVendorText('{\n  "functionCall": {')).toBe(true);
    expect(isHiddenVendorText('Got it.')).toBe(false);
  });
});
